implementations/elixir/ockam/ockam/lib/ockam/transport/portal/inlet_listener.ex

```diff
@@ -0,0 +1,49 @@
+defmodule Ockam.Transport.Portal.InletListener do
+    @moduledoc """
+    GenServer implementing the Inlet TCP listener 
+    It's a GenServer just to make it simple to be added
+    to a supervision tree. 
+    """
+
+    require Logger
+    use GenServer
+
+    @typedoc """
+    TCP listener options
+    - ip: t::inet.ip_address() - IP address to listen on
+    - port: t:integer() - port to listen on
+    - peer_route: route to outlet
+    """
+    @type options :: Keyword.t()
+
+    def start_link(options) do
+      GenServer.start_link(__MODULE__, options)
+    end
+
+    @impl true
+    def init(options) do
+      ip = Keyword.get_lazy(options, :ip, &default_ip/0)
+      port = Keyword.get_lazy(options, :port, &default_port/0)
+      peer_route = options[:peer_route]
+      Logger.info("Starting inlet listener on #{inspect(ip)}:#{port}. Peer: #{inspect(peer_route)}")
+      {:ok, lsocket} = :gen_tcp.listen(port, [:binary, {:active, false}, {:ip, ip}, {:reuseaddr, true}])
+      spawn_link(fn -> accept(lsocket, peer_route) end)
+      {:ok, %{listen_socket: lsocket, peer_route: peer_route}}
+    end
+
+    defp accept(lsocket, peer_route) do
+      {:ok, socket} = :gen_tcp.accept(lsocket)
+      {:ok, worker} = Ockam.Transport.Portal.InletWorker.create([peer_route: peer_route])
+      case Ockam.Node.whereis(worker) do
+        nil ->
+          raise "Worker #{inspect(worker)} not found"
+        pid ->
+          :ok =:gen_tcp.controlling_process(socket, pid)
+          GenServer.cast(pid, {:takeover, socket})
```

Comment:
We can pass the socket when creating the InletWorker,  but we would need anyway some form of synchronization to tell the worker it's safe to set it to active mode.  We can't set it to active mode *until* we have transferred the ownership of the socket,  otherwise risking tcp' data to be delivered to the parent (listening) process instead of the worker.
Per [gen_tcp:controlling_process/2 ](https://www.erlang.org/doc/man/gen_tcp.html#controlling_process-2)docs: 

> For instance changing the sockets active mode before the transfer is complete may cause this.

Being realistic,  it's true true that it should be _very_ remote possibility of this to happens, particularly because the worker doesn't set the socket to active until receiving the :pong from the other side.